### PR TITLE
many: hook in start of code to fetch/check assertions when installing snap from store

### PR DESCRIPTION
**Pull request by pedronis**
After some thinking it seems InstallPath will be better served by a fail early approach (treating there checking assertions like we treat talking to the store in other cases) so it seems a fetch and check task seems appropriate to cover the install from store needs (until update control enters the picture at least).

### CODE REVIEW

**Comment by pedronis** [comment]
rebased on #1713

**Comment by pedronis** [comment]
retest this please

**Comment by pedronis** [comment]
retest this please

File: store/store.go [hunk 1]
```diff
@@ -1018,7 +1018,8 @@ func (s *Store) Assertion(assertType *asserts.AssertionType, primaryKey []string
 	defer resp.Body.Close()
 
 	if resp.StatusCode != 200 {
-		if resp.Header.Get("Content-Type") == "application/json" {
+		contentType := resp.Header.Get("Content-Type")
+		if contentType == "application/json" || contentType == "application/problem+json" {
```
Comment (mvo5) [inline store/store.go]: Why this change? Just curious, seems unreleated to the rest.

Comment (pedronis) [inline store/store.go]: @mvo5 because I saw I was getting the wrong kind of errors when the spread tests were failing, SAS is returning problem+json not just json

File: overlord/assertstate/assertmgr.go [hunk 2]
```diff
@@ -110,6 +113,14 @@ func userFromUserID(st *state.State, userID int) (*auth.UserState, error) {
 	return auth.User(st, userID)
 }
 
+type assertionNotFoundError struct {
+	ref *asserts.Ref
+}
+
+func (e *assertionNotFoundError) Error() string {
+	return fmt.Sprintf("%s %v not found", e.ref.Type.Name, e.ref.PrimaryKey)
```
Comment (mvo5) [inline overlord/assertstate/assertmgr.go]: (nitpick) I guess we could use `e.ref.Unique()` here to make it a bit more copy/paste friendly.

Comment (pedronis) [inline overlord/assertstate/assertmgr.go]: @mvo5 I have been actually thinking of giving a String method to Ref, just couldn't commit to how it should look like

**Comment by mvo5** [comment]
:+1:

File: overlord/assertstate/assertmgr.go (hunk 2 above)
Comment (mvo5) [inline overlord/assertstate/assertmgr.go]: :)

File: overlord/assertstate/assertmgr.go [hunk 3]
```diff
@@ -45,8 +49,7 @@ type AssertManager struct {
 func Manager(s *state.State) (*AssertManager, error) {
 	runner := state.NewTaskRunner(s)
 
-	runner.AddHandler("fetch-snap-assertions", doFetchSnapAssertions, undoFetchSnapAssertions)
-	// TODO: check-snap-assertions handlers
+	runner.AddHandler("fetch-check-snap-assertions", doFetchCheckSnapAssertions, undoFetchCheckSnapAssertions)
```
Comment (niemeyer) [inline overlord/assertstate/assertmgr.go]: Can we use `"validate-snap"` here?  It's shorter, seems a bit more inline with the remaining snap tasks, it's quite "legally binding", and it takes a small step back from the nitty-gritty of the operation, meaning some freedom to operate internally.

File: overlord/assertstate/assertmgr.go [hunk 4]
```diff
@@ -189,10 +202,79 @@ func doFetchSnapAssertions(t *state.Task, _ *tomb.Tomb) error {
 		PrimaryKey: []string{sha3_384},
 	}
 
-	return fetch(t.State(), ref, ss.UserID)
+	err = fetch(t.State(), ref, ss.UserID)
+	if notFound, ok := err.(*assertionNotFoundError); ok {
+		if notFound.ref.Type == asserts.SnapRevisionType {
+			return fmt.Errorf("cannot verify snap %q and its hash, no matching assertions found", ss.Name())
```
Comment (niemeyer) [inline overlord/assertstate/assertmgr.go]: I suggest s/and its hash// here and below. The hash _is_ the snap.

Also suggest replacing the word "assertions" here and below by "signatures". Although a bit less precise for us, it's not completely wrong and it's a much more common term in this context.

File: overlord/assertstate/assertmgr.go [hunk 5]
```diff
@@ -189,10 +202,79 @@ func doFetchSnapAssertions(t *state.Task, _ *tomb.Tomb) error {
...
+	if err != nil {
+		return err
+	}
+
+	// TODO: set DeveloperID from assertions
+	return nil
 }
 
-func undoFetchSnapAssertions(t *state.Task, _ *tomb.Tomb) error {
+func undoFetchCheckSnapAssertions(t *state.Task, _ *tomb.Tomb) error {
```
Comment (niemeyer) [inline overlord/assertstate/assertmgr.go]: undo is still optional isn't it?  We can just kill this function I think.

File: overlord/assertstate/assertmgr.go [hunk 6]
```diff
@@ -189,10 +202,79 @@ func doFetchSnapAssertions(t *state.Task, _ *tomb.Tomb) error {
...
 	// nothing to do, the assertions that were *actually* added are still true
 	return nil
 }
+
+func crossCheckSnap(st *state.State, name, snapSHA3_384 string, snapSize uint64, si *snap.SideInfo) error {
+	// get relevant assertions and and do cross checks
```
Comment (niemeyer) [inline overlord/assertstate/assertmgr.go]: and and

File: overlord/assertstate/assertmgr.go [hunk 7]
```diff
@@ -189,10 +202,79 @@ func doFetchSnapAssertions(t *state.Task, _ *tomb.Tomb) error {
...
+	a, err := db.Find(asserts.SnapRevisionType, map[string]string{
+		"snap-sha3-384": snapSHA3_384,
+	})
+	if err != nil {
+		return fmt.Errorf("internal error: cannot find just fetched snap-revision assertion for %q: %s", name, snapSHA3_384)
+	}
+	snapRev := a.(*asserts.SnapRevision)
+
+	if snapRev.SnapSize() != snapSize {
+		return fmt.Errorf("snap %q file does not have expected size according to assertions (download is broken or tampered): %d != %d", name, snapSize, snapRev.SnapSize())
```
Comment (niemeyer) [inline overlord/assertstate/assertmgr.go]: Finally! :)

File: overlord/assertstate/assertmgr.go [hunk 8]
```diff
@@ -189,10 +202,79 @@ func doFetchSnapAssertions(t *state.Task, _ *tomb.Tomb) error {
...
+
+	snapID := si.SnapID
+
+	if snapRev.SnapID() != snapID || snapRev.SnapRevision() != si.Revision.N {
+		// we have at least 2 cases here, what's the best message?
+		// - an unsuccesufl MITM
+		// - broken store metadata resulting into broken assertions
+		//   (more likely if it is snap-revision not matching)
+		//   people would need to report this
+		return fmt.Errorf("snap %q file hash %q corresponding assertions implied snap id %q and revision %d are not the ones expected for installing (store metadata is broken or communication tampered): %q and %s", name, snapSHA3_384, snapRev.SnapID(), snapRev.SnapRevision(), snapID, si.Revision)
```
Comment (niemeyer) [inline overlord/assertstate/assertmgr.go]: I had a great feeling when reading the message above. Feels simple and to the point. Can we mimic it here? This one is a bit hard to go through. E.g.:

`"snap %q does not have expected ID or revision according to assertions (metadata is broken or tampered): %d / %s != %d / %s"`

File: overlord/assertstate/assertmgr.go [hunk 9]
```diff
@@ -189,10 +202,79 @@ func doFetchSnapAssertions(t *state.Task, _ *tomb.Tomb) error {
...
+		"series":  release.Series,
+		"snap-id": snapID,
+	})
+	if err != nil {
+		return fmt.Errorf("internal error: cannot find just fetched snap declaration for %q: %s", name, snapID)
+	}
+	snapDecl := a.(*asserts.SnapDeclaration)
+
+	if snapDecl.SnapName() == "" {
+		// TODO: trigger a global sanity check
```
Comment (niemeyer) [inline overlord/assertstate/assertmgr.go]: Not sure we need those triggers. The system should be able to self-correct over time nevertheless, so we might just wait for the next pass and have a single code path instead of worrying here.

File: overlord/managers_test.go [hunk 10]
```diff
@@ -55,6 +61,9 @@ type mgrsSuite struct {
 	udev       *testutil.MockCmd
 	prevctlCmd func(...string) ([]byte, error)
 
+	storeSigning *assertstest.StoreStack
+	restore      func()
```
Comment (niemeyer) [inline overlord/managers_test.go]: Restore what?  It's also not clear why this isn't just using defer as usual.

File: overlord/managers_test.go [hunk 11]
```diff
@@ -78,13 +87,19 @@ func (ms *mgrsSuite) SetUpTest(c *C) {
 	ms.aa = testutil.MockCommand(c, "apparmor_parser", "")
 	ms.udev = testutil.MockCommand(c, "udevadm", "")
 
+	rootPrivKey, _ := assertstest.GenerateKey(1024)
+	storePrivKey, _ := assertstest.GenerateKey(752)
```
Comment (niemeyer) [inline overlord/managers_test.go]: Assert(err, IsNil) on both?

**Comment by niemeyer** [comment]
LGTM when you're happy about these comments.

File: overlord/assertstate/assertmgr.go (hunk 9 above)
Comment (pedronis) [inline overlord/assertstate/assertmgr.go]: yes, we have some options here, I agree/plan to have regular passes (basicall code in AssertManager.Ensure I think atm), just a bit unsure how often we can do them, so this would be more about timeliness than correctness

File: overlord/managers_test.go (hunk 11 above)
Comment (pedronis) [inline overlord/managers_test.go]: this test-only helpers actually panic (it's the style of the fully asserttest package), this is just ignoring a 2nd result that we don't use here

File: overlord/assertstate/assertmgr.go (hunk 5 above)
Comment (pedronis) [inline overlord/assertstate/assertmgr.go]: I'm always a bit confused on this matter, I'll check this again
